File: 2016/go/day07/day07_test.go
package day07

import (
	"reflect"
	"testing"
)

func TestCountIPsSupportingTLS(t *testing.T) {
	type args struct {
		str []string
	}
	tests := []struct {
		name string
		args args
		want int
	}{
		{"1", args{[]string{
			"abba[mnop]qrst",
			"abcd[bddb]xyyx",
			"aaaa[qwer]tyui",
			"ioxxoj[asdfgh]zxcvbn",
			}}, 2},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := CountIPsSupportingTLS(tt.args.str); got != tt.want {
				t.Errorf("CountIPsSupportingTLS() = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_parseLines(t *testing.T) {
	type args struct {
		str []string
	}
	tests := []struct {
		name string
		args args
		want []Ip
	}{
		{"1", args{[]string{"abba[mnop]qrst"}}, []Ip{{"abba|qrst", "mnop"}}},
		{"2", args{[]string{"abcd[bddb]xyyx"}}, []Ip{{"abcd|xyyx", "bddb"}}},
		{"3", args{[]string{"aaaa[qwer]tyui"}}, []Ip{{"aaaa|tyui", "qwer"}}},
		{"4", args{[]string{"ioxxoj[asdfgh]zxcvbn"}}, []Ip{{"ioxxoj|zxcvbn", "asdfgh"}}},
		{"4", args{[]string{"ioxxoj[asdfgh]zxcvbn", "aaaa[qwer]tyui"}}, []Ip{{"ioxxoj|zxcvbn", "asdfgh"}, {"aaaa|tyui", "qwer"}}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := parseLines(tt.args.str); !reflect.DeepEqual(got, tt.want) {
				t.Errorf("parseLines() = %v, want %v", got, tt.want)
			}
		})
	}
}

File: 2016/ts/test/day07.test.ts
import {Ip, Snooping} from '../src/day07'

describe('IP should', () => {
    it('detect ABBA in "abba"', ()=>{
        let i = new Ip( 'abba' );

        expect(i.supportsTLS()).toBe(true)
    })
    it('detect invalid ABBA in "abcd[bddb]xyyx"', ()=>{
        let i = new Ip( 'abcd[bddb]xyyx' );

        expect(i.supportsTLS()).toBe(false)
    })
    it('detect invalid ABBA in "aaaa[qwer]tyui"', ()=>{
        let i = new Ip( 'aaaa[qwer]tyui' );

        expect(i.supportsTLS()).toBe(false)
    })
    it('detect ABBA in "ioxxoj[asdfgh]zxcvbn"', ()=>{
        let i = new Ip( 'ioxxoj[asdfgh]zxcvbn' );

        expect(i.supportsTLS()).toBe(true)
    })
    it('detect ABA in "aba[bab]xyz"', ()=>{
        let i = new Ip( 'aba[bab]xyz' );

        expect(i.supportsSSL()).toBe(true)
    })
    it('detect invalid ABA in "xyx[xyx]xyx"', ()=>{
        let i = new Ip( 'xyx[xyx]xyx' );

        expect(i.supportsSSL()).toBe(false)
    })
    it('detect ABA in "aaa[kek]eke"', ()=>{
        let i = new Ip( 'aaa[kek]eke' );

        expect(i.supportsSSL()).toBe(true)
    })

    it('detect ABA in "zazbz[bzb]cdb"', ()=>{
        let i = new Ip( 'zazbz[bzb]cdb' );

        expect(i.supportsSSL()).toBe(true)
    })

    it('detect ABA in "zazaza"', ()=>{
        let i = new Ip( 'zazaz' );

        expect(i.supportsSSL()).toBe(false)
    })
    it('detect ABA in "zaz[aba]aza"', ()=>{
        let i = new Ip( 'zaz[aba]aza' );

        expect(i.supportsSSL()).toBe(false)
    })

})

describe('Snooping should', () => {
    it('should return answer', () => {

        let i = new Snooping()
        expect(i.Part1()).toBe(115)
    })
    it('should return answer', () => {

        let i = new Snooping()
        expect(i.Part2()).toBe(231)
    })
})
